beer-garden/functionary
# Refs #152 - Add third party auth support

This PR integrates `django-allauth` for supporting oauth2 and openid providers. This allows additional oauth providers to be configured for authentication in addition to the standard login/password.

When signing in with a third party auth, a User object will be created without access to any Functionary environments. When logging in without any environment access, a message will be displayed indicating this and to contact an admin to get access to environments instead of just showing a bunch of blank pages.

Clicking on the user name in the Functionary UI will display associated third party accounts and allow for disconnecting from the provider.

## Testing
To enable a third party auth via gitlab make the following modifications, adjusting `localhost` for your actual hostname.

Add the following line to the `settings/__init__.py`:
```
    "allauth.socialaccount.providers.gitlab",
```
Also, add the following to `settings/auth_.py`:
```python
SOCIALACCOUNT_PROVIDERS = {
    "gitlab": {
        "GITLAB_URL": "https://<gitlab url>",
    },
}
```

Go to *Gitlab* -> *User Settings* -> *Application*
Enter the following:
  - Name: "Functionary Auth API"
  - Redirect URI: "http://localhost:8000/accounts/gitlab/login/callback/"
  - Confidential: Checked
  - Scopes: read_user

Copy the `Application ID` and `Secret`
Log into the Django Admin
Add a *Site*: 
  - Set the `Domain Name` to the FQDN or the IP of the Django server
  - Set the `Display Name` to something meaningful.
  - 
Add a *Social Application*:
  - Select `GitLab` for the *Provider*
  - Set the *Name* to something meaningful
  - Set the *Client Id* to the `Application ID` from Gitlab
  - Set the *Secret Key* to the `Secret` from Gitlab
  - Select the *Site* created in the previous step and click the right pointing arrow to add it
Click `Save` to add the application.

You may need to set the following when launching Functionary if you get certificate errors when using a third party provider. If this is the case, make sure the certificate chain is correct, such as by setting the environment with:
```bash
REQUESTS_CA_BUNDLE = "/etc/ssl/certs/ca-certificates.crt"
```
